Assets/Scripts/Effects/EffectSpawner.cs

```diff
@@ -20,8 +20,19 @@ public class EffectSpawner : MonoBehaviour
     private void Start()
     {
         stateManager = this.EnsureComponent<PlayerStateManager>();
-        stateManager.CallOnStateEnter(triggerState, StateStart);
-        stateManager.CallOnStateExit(triggerState, StateEnd);
+        stateManager.OnStateChange += HandleNewPlayerState;
+    }
+
+    private void HandleNewPlayerState(State oldState, State newState)
+    {
+        if (newState == triggerState)
+        {
+            StateStart();
+        }
+        if (oldState == triggerState)
```

Comment:
Why not `else if`? Would you ever have a transition someState -> someState?

Reply:
No strong reason in particular other than I am conceptually separating out how I think about it. 1) Handle new state 2) Handle cleanup from old state